src/g_kbdnav.h

```diff
@@ -0,0 +1,165 @@
+#ifndef HAVE_KEYBOARDNAV
+        #define HAVE_KEYBOARDNAV 1
+#endif
+
+#ifndef G_KBDNAV_H
+#define G_KBDNAV_H
+
+typedef struct _kbdnav
+{
+    int kn_state;        /* see KN_ consts below */
+    int kn_ioindex;      /* selected inlet/outlet, 0-based*/
+    int kn_moddown;      /* is our modifier key down? */
+    int kn_connindex;    /* selected in/out connection, 0-based */
+    t_outconnect *kn_outconnect;
+    t_object *kn_selobj; /* currently selected object */
+    char* kn_highlight;  /* highlight_color */
+    struct _linetraverser *kn_linetraverser;
+    struct _magtrav *kn_magtrav; /* linked list of connections for the magnetic connector*/
+    int kn_waitingfornumber;
+    int kn_chosennumber; /* used by the digit connector */
+    int kn_iotype;       /* IO_INLET/IO_OUTLET */
+    int kn_indexvis;     /* boolean: should we visualize object indexes? */
+} t_kbdnav;
+
+/* #define constants */
+
+#define KN_INACTIVE 0 /* not navigating with the keyboard */
+#define KN_IO_SELECTED 1 /* selecting an inlet/outlet */
+#define KN_CONN_SELECTED 2 /* navigating through an in/outlet's connections */
+#define KN_WAITING_NUMBER 3 /* waiting for a number from 0-9 when using the "digit connector" */
+
+#define KBDNAV_SEL_PAD_X 4
+#define KBDNAV_SEL_PAD_Y 3
+
+#define IO_INLET 0
+#define IO_OUTLET 1
+
+/* FORWARD DECLARATIONS */
+
+/* stuff from g_editor.c */
+
+EXTERN void canvas_connect_with_undo(t_canvas *x, t_float index1, t_float outno,
```

Comment:
these functions shouldn't be declared as `EXTERN`: this is a private API and we don't want externals to use it.

Reply:
Okay. 

I recall that i tought i would need to expose this function so the `connect_with_undo` message sent from the gui could work. 

But now that i think about it i wonder: i actually can send a `connect_with_undo` from the guibut only doing [this class_addmethod() call](https://github.com/pure-data/pure-data/pull/869/files#r368252839) without removing the "static" from the function?

Of course as you said in the comment it still would expose 'connect_with_undo' to the user so one shouldn't do it. I'm just asking to better understand how PD (and C) works.

Reply:
Anyway, when implementing a dedicated embbeded "connect" dialog i would need to somehow make the gui ask for the core to make a connection. How should it be done?

Just make the gui call a (non-exposed) function - lets say, for the sake of clarity "receive_connection_request_from_gui()" -that in turn would call `canvas_connect_with_undo`?

Also, this function would need to be located in `g_editor.c` (since `canvas_connect_with_undo` would be static), right?

Reply:
fixed e998b865392d8c169b9b784f05b0d2c8f1eb6ea7